### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by jbednar**
I've made a pass on the ParamNB code, coming up with an implementation that I think is reasonably clean.  

First, I changed the newly defined Parameter types (FileSelector and MultiFileSelector) to be independent of ParamNB, with full implementations that make sense whether or not they are used in the context of a notebook.  This fits with the spirit of Param, and allows those definitions to be moved into Param itself, where they belong (though they are currently still in this library for the purposes of review).  They are now just variants of Param.ObjectSelector.

Second, now that those Parameter types implement their own logic, I was able to remove all the special-case handling for specific Parameter types in _make_widget.  That method no longer needs to check for specific subclasses, and instead handles everything via a lookup table that matches Parameter types to ipywidget types.  I added a dictionary of options to this lookup table, in case people want to set ipywidget options per Parameter type, and used the existing example of width for MultiFileSelector, but I'm not certain this is necessary.

I also cleaned up the code in general.

Remaining issues:
- I can't tell what "Halt" achieves; it doesn't seem to do much whether I enable it or not.  What I had expected was that if enabled and I ran the notebook with `Run All`, only the cells up to and including that cell would be executed.  That's not what happens.
- I'm not entirely sure how the `execute` option is meant to be used in general; it doesn't seem clear what control is being provided.
- Maybe instead of `next` the option should be an integer, to run n numbers of subsequent cells?  Otherwise one can't put a markdown box in between the cell and the subsequent cell.
- In another project, we provided persistent state via pickling, which is awkward.  Can we provide persistent state using the notebook's metadata?
- The print_param_defaults function in Param has broken formatting, perhaps because it needs to be updated for modern Python print function syntax. If we move the new Parameter types to Param then we should probably fix that.
- Does anyone else see other issues?

### CODE REVIEW

**Comment by philippjfr** [comment]
> I added a dictionary of options to this lookup table, in case people want to set ipywidget options per Parameter type, and used the existing example of width for MultiFileSelector, but I'm not certain this is necessary.

I think it's a good idea anyway, I'm sure there will be some widget or parameter type that will require special handling.

> I can't tell what "Halt" achieves; it doesn't seem to do much whether I enable it or not. What I had expected was that if enabled and I ran the notebook with Run All, only the cells up to and including that cell would be executed. That's not what happens.

I think the code to do with execution and halt are a bit muddled up. I'll have to look at that.

> I'm not entirely sure how the execute option is meant to be used in general; it doesn't seem clear what control is being provided.

As you suggest it might be better to make this an integer to specify the number of subsequent cells to run.

> Otherwise one can't put a markdown box in between the cell and the subsequent cell.

Markdown cells don't count, but running 'below' has issues and next is too limited.

> In another project, we provided persistent state via pickling, which is awkward. Can we provide persistent state using the notebook's metadata?

Should be possible. 

It looks good to me and is what I expected after our discussion but we should wait on @marqh to respond because this will require some changes to scitools/CubeBrowser.

**Comment by jlstevens** [comment]
> Markdown cells don't count, but running 'below' has issues and next is too limited.

Just to mention that in the IPython Javascript API there is an option to execute a cell range. Seems like functionality worth exposing to me for this sort of reason.

**Edit 1:** We are already using it, but to execute one step `nb.execute_cell_range(index+1, index+2);`

**Edit 2:** You can even imagine supplying a list of tuples to execute a range, skip some cells, execute a later range etc. Not sure it is useful though.

**Edit 3:** Actually, being able to execute cell ranges that are _before_ the current widget cell seems like it could be very useful, opening up new ways of controlling execution via widgets.

**Comment by philippjfr** [comment]
I think it should be straightforward to think about and since its mostly about chaining, executing the next n cells is what I had in mind and is also what Jim suggested. I can't quite see what running previous cells would accomplish and it seems confusing and weird to me, as does supplying lists or any other subset.

Edit: I can see why it would be useful I just don't think it fits into this widget framework. Also note that the specification here will be relative to the current cell, which is why I'm hesitant to support subsets or custom ranges, which I think would be a recipe for confusion, if you move even just one cell around.

**Comment by jlstevens** [comment]
I agree it would be fragile. I only mentioned this as an idea for an additional option...not the default way of doing things. I'm just saying we could expose more control _if_ the users want to use it.

**Comment by philippjfr** [comment]
Well the point here was to decide on a sensible replacement for the current execute option. Replacing it with an option to execute the next N cells is very straightforward and allows chaining a bunch of widgets and plots together easily. 

Currently you can register whatever JS hook you like, so it already offers that freedom. But I don't actually think that freedom is desirable in this case, since it will lead you down a path where the order of execution will be very hard to reason about. 

Edit: Very possible I'm ignoring an important use case where some custom cell execution model would be required though.

**Comment by jbednar** [comment]
I agree with @philippjfr -- providing arbitrary out-of-order execution could indeed be useful in some cases, but it would also be extraordinarily confusing, and so I think we should work with "run next n cells" and "run all cells below" until we come up with a very compelling reason we need more power than that.  

With that in mind, can we agree on the general principles so that it's clear what to fix (if anything) in how it behaves?  I think:
1. By default, Widgets() should simply carry on to the next cell just like any normal object would ever do.  That way, if users are running the notebook one cell at a time, they would interactively adjust the values before moving on to the next cell, and if they do "Run All", they'll have no chance to change values, but they call always edit the values then do "Run All Below" from Jupyter to have the new values affect later cells.  I think this is the least surprising default behavior.
2. If `halt=True`,  it works just the same as in 1., except that for Run All it doesn't actually run everything, but instead stops after this cell to force the user to be able to edit the parameters before moving on.  I _think_ that was the intent of `halt`, but I'm not certain, since that's not what it appears to do now.  By itself, doesn't add any special buttons; users would just do "Run All Below" to continue afterwards, or run the cells one at a time.
3. If `execute='below'`, provide a `Run below` button that is equivalent to Jupyter's "Run All Below" selected when that cell is active.  The button is simply a convenience, and doesn't change the behavior in any way until it is pressed, at which point it's the same as if the menu option had been selected at that point.  In particular, halting would only happen if `halt==True`.
4. If `execute=2` (where 2 can be any positive int), provide a `Next n` button (with n shown if greater than 1, perhaps), which again does not change the behavior in any way, i.e. not halting, and is simply a convenience that runs the next n cells when pushed.
5. if specified, a callback will be called whenever a parameter value changes (if `onchange==True`) or whenever one of these new buttons is pressed (if `onchange==False`).

Does this seem like a reasonable design?  Probably `execute` would need a new name, as it's really just about making a button rather than controlling execution, in this approach.  I think the above description is quite different from what's actually implemented, but to me it seems much easier to understand and to use than whatever it's currently doing.  Yet because I don't totally understand what it's currently doing and why, I'm not completely confident that this design is appropriate.  Feedback?  E.g. @ceball?

**Comment by philippjfr** [comment]
> By default, Widgets() should simply carry on to the next cell just like any normal object would ever do. That way, if users are running the notebook one cell at a time, they would interactively adjust the values before moving on to the next cell, and if they do "Run All", they'll have no chance to change values, but they call always edit the values then do "Run All Below" from Jupyter to have the new values affect later cells. I think this is the least surprising default behavior.

Agreed.

> If halt=True, it works just the same as above, except that for Run All it doesn't actually run everything, but instead stops after this cell to force the user to be able to edit the parameters before moving on. I think that was the intent of halt, but I'm not certain, since that's not what it appears to do now. By itself, doesn't add any special buttons; users would just do "Run All Below" to continue afterwards, or run the cells one at a time.

This sounds sensible. However there is an issue with "Run All Below". Say you have two separate cells that have a "Run All Below" button, you click the first, it proceeds to the second one and halts. So far so good, but if you now click "Run All Below" on the second widget you will have scheduled the execution of all the cells below twice. Therefore I think "Run next n cells" is the only useful and non-confusing option.

> If execute='below', provide a Run below button that is equivalent to Jupyter's "Run All Below" selected when that cell is active. The button is simply a convenience, and doesn't change the behavior in any way. E.g. halting would only happen if halt==True.

That sounds sensible and is what I think I had originally intended but I muddled up the `halt` and `execute` logic.

> If execute=2 (where 2 can be any positive int), provide a Next n button (with n shown if greater than 1, perhaps), which again does not change the behavior in any way, i.e. not halting, and is simply a convenience that runs the next n cells when pushed.

Sounds good.

> if specified, a callback will be called whenever a parameter value changes (if on_change==True) or whenever one of these new buttons is pressed (if on_change==False).

Sounds good. One question though, currently `onchange=True` will replace the execute button. Would you eliminate that? Also if `onchange=False` would the `callback` also still get called on a button press?

**Comment by jbednar** [comment]
> However there is an issue with "Run All Below". Say you have two separate cells that have a "Run All Below" button, you click the first, it proceeds to the second one and halts. So far so good, but if you now click "Run All Below" on the second widget you will have scheduled the execution of all the cells below twice. Therefore I think "Run next n cells" is the only useful and non-confusing option.

Are you saying that this is a design problem, or an implementation problem?  From a design perspective, I don't see any problem with "Run all below"; it should run all the cells below this one and halt if it encounters one that says halt; seems very well defined and consistent.  Are you saying that, unfortunately, the implementation doesn't allow that, such that instead of doing something reasonable it has to schedule every cell to run, in a way that the post-halt cells aren't dropped from the execution list?  From the code I can see what you mean, but is there not some way to indicate "start running from this cell" without saying "run from cell x to cell n"?  If there is no way to express that in the current Jupyter API, then I agree that "Run all below" is too tricky and confusing.  But again, from a design perspective, it seems like a reasonable behavior to _try_ to do, if it's feasible.

> One question though, currently onchange=True will replace the execute button. Would you eliminate that? Also if onchange=False would the callback also still get called on a button press?

In general I strongly prefer independent, combinatorial options, each specifying one aspect of the behavior and not messing with any other aspects.  In this case I would not expect onchange to affect any of the buttons, but I'm not completely sure how to best use callbacks at all, so don't trust me too much.  I do think that if onchange==False, the callback would get called on a button press, and indeed that's the only time it would get called in that case (right?).

**Comment by jbednar** [comment]
Some of this behavior is also related to persistence -- if we support persistence using the metadata associated with the cell, then the user can edit the values and re-run the entire notebook, allowing them to make changes and see the results even without running a cell at a time or using the buttons discussed above.  That doesn't solve _everything_, but maybe makes it less important to provide lots of execution-control features.

**Comment by philippjfr** [comment]
> Are you saying that, unfortunately, the implementation doesn't allow that, such that instead of doing something reasonable it has to schedule every cell to run, in a way that the post-halt cells aren't dropped from the execution list?

Yes, that's what I'm saying. I agree that conceptually it's a very clear model and if it could be made to work the way you described would probably be the most sensible default behavior. Unfortunately once the notebook has scheduled a cell range to run I don't know of anyway of deduplicating the scheduled events, i.e. the second widget in my example has no way of knowing if the cells below it have already been scheduled to run.

> I do think that if onchange==False, the callback would get called on a button press, and indeed that's the only time it would get called in that case (right?).

Correct. I do think `onchange` is a combinatorial option, it basically says execute on button press OR execute when on the widgets changes state. However both the button press and the onchange True behavior execute both the callback AND the execution hook (i.e. which cells to execute). You're right that if no execution and no callback is defined then onchange does nothing at all, so I guess it's not completely combinatorial.

**Comment by jbednar** [comment]
> Yes, that's what I'm saying. I agree that conceptually it's a very clear model and if it could be made to work the way you described would probably be the most sensible default behavior. Unfortunately once the notebook has scheduled a cell range to run I don't know of anyway of deduplicating the scheduled events, i.e. the second widget in my example has no way of knowing if the cells below it have already been scheduled to run.

Ok.  On the other hand, it's probably quite common to have some parameters at the top of the file separated from the rest of the code by a barrier, and the current "Run Below" would work perfectly for that (i.e., as long as it's the only such button on the page).  Whereas I don't see any way for 'Run Next n' to work reasonably in that case, or at least no more reasonably than "Below" would (e.g. if n=num_remaining_cells).   So I think it still makes sense to offer "Run Below", as long as we say there can be only one of those in the notebook.

> Correct. I do think onchange is a combinatorial option, it basically says execute on button press OR execute when on the widgets changes state. However both the button press and the onchange True behavior execute both the callback AND the execution hook (i.e. which cells to execute). You're right that if no execution and no callback is defined then onchange does nothing at all, so I guess it's not completely combinatorial.

Ah, I hadn't gotten that nuance. So if `onchange==True`, there will be no button, and instead the 'next n' or all cells below will be executed, whenever a widget is dragged?  Ooh, that's tricky, but I see why it could be cool -- have params in one cell, and have them directly control the next cell or two.  In that case there really should have no buttons appearing, right?

**Comment by philippjfr** [comment]
> Ok. On the other hand, it's probably quite common to have some parameters at the top of the file separated from the rest of the code by a barrier, and the current "Run Below" would work perfectly for that (i.e., as long as it's the only such button on the page). Whereas I don't see any way for 'Run Next n' to work reasonably in that case, or at least no more reasonably than "Below" would (e.g. if n=num_remaining_cells). So I think it still makes sense to offer "Run Below", as long as we say there can be only one of those in the notebook.

Okay I agree with that, as long as we're clear about it. However run all below is really just a special case of running the next N cells, if we name the parameter correctly then we could just reserve a keyword like 'all' to run all cells below.

> Ah, I hadn't gotten that nuance. So if onchange==True, there will be no button, and instead the 'next n' or all cells below will be executed, whenever a widget is dragged? Ooh, that's tricky, but I see why it could be cool -- have params in one cell, and have them directly control the next cell or two. In that case there really should have no buttons appearing, right?

That's right, I think it is a pretty cool feature. Maybe we can invert it though and make it about giving you a choice between button or no button, would that be clearer? It's perhaps clearer that if you enable a button but don't tell it to execute anything (cells or callback) that you won't actually get a button. Or it'll warn if you say button=True but you haven't actually told it to execute anything.

In other words defining `button=True, execute=None, callback=None` and getting a warning that the button wasn't added because it would do nothing would be clearer than specifying `onchange=False, execute=None, callback=None` and then getting a warning about buttons.

Edit: Thinking about that more I'm unsure again, a button option would probably be True by default, in which case it would warn whenever you specify no execution or callback.

**Comment by jbednar** [comment]
> Okay I agree with that, as long as we're clear about it. However run all below is really just a special case of running the next N cells, if we name the parameter correctly then we could just reserve a keyword like 'all' to run all cells below.

Good point; I agree.

> a button option would probably be True by default, in which case it would warn whenever you specify no execution or callback.

I can't quite wrap my head around these other options.  I don't think a button should be provided by default, because I don't think we should be allowing the interactive user to mess with the notebook's machinery unless the notebook author explicitly requested that on Widgets instantiation.

Maybe the options would be `next=0,1,2,...,'all'`, `button=False,True`, `callback=None,fn_name`, where the first one is the default in each case?  In every case, `next` is the number of subsequent cells to execute when executing any such cells, which is pretty clear.  The other cases have to be enumerated:

button=False, callback=None: run next n cells, whever a widget is dragged
button=True, callback=None: run next n cells, when the button is pushed
button=False, callback=fn_name: run next n cells _and_ run fn_name, whenever a widget is dragged
button=True, callback=fn_name: run next n cells _and_ run fn_name, when the button is pushed

Does that sound right?  Does it cover all the possibilities?

**Comment by philippjfr** [comment]
Yes, I think we've finally got it. That all makes sense to me and seems pretty clear.

**Comment by jbednar** [comment]
Great.  I've edited the descriptions above slightly to make them be more clearly combinatorial.  Can you double-check them one last time?

**Comment by philippjfr** [comment]
Still looks good. Will you need me to look at the `hold` logic?

**Comment by jbednar** [comment]
`halt`?  I guess that's another combinatorial option.  I'm not sure whether I'll be able to figure out how to implement _any_ of this :-), but I can try (on Monday).  At least we now have a design to work towards.

**Comment by jbednar** [comment]
On another topic, the way I've implemented FileSelector and MultiFileSelector is only halfway to where they need to be.  Specifically, they don't currently implement a way for the user to edit the glob path itself.  In the other version of paramnb that we're using in a different project, there were complete hacks to do that using two separate parameters, but to me it seems like those ought to be two separate _widgets_ that appear for the same parameter.  I.e., right now FileSelector shows only a dropdown, but it should presumably show a dropdown for selecting files _and_ a text box for typing a glob/path.  Does ipywidgets have any reasonable way to make a widget out of a pair of widgets to handle this situation?  I'm not sure what to do otherwise; I certainly don't want to do what the other paramnb version was doing.

**Comment by philippjfr** [comment]
One last bit to discuss is the behavior of the callback. Currently it passes the class or instance to the callback so it can easily access the parameters. I'd also like to detect if the callback is a member of the object and not pass in the object in that case because it will already be available on self.

I know you didn't like the pattern of adding a callback on the class itself but I think it's very useful. It will let you encapsulate a GUI on a single class.

**Comment by jbednar** [comment]
Yes, it would make sense to suppress passing the object if it would already be available via self.

The mechanism for running the next n cells on every change seems like it's a very notebook-y way to do a GUI that  essentially treats the next cell(s) as the callback, but sure, in other cases one won't want to be so tied to the notebook, and instead will just want one cell that does everything, instantiated easily, in which case the callback will be useful.
